libs/vkd3d/resource.c

```diff
@@ -3048,18 +3071,28 @@ static void vkd3d_create_buffer_uav(struct d3d12_desc *descriptor, struct d3d12_
 
     if (counter_resource)
     {
-        const struct vkd3d_format *format;
-
         assert(d3d12_resource_is_buffer(counter_resource));
         assert(desc->u.Buffer.StructureByteStride);
 
-        format = vkd3d_get_format(device, DXGI_FORMAT_R32_UINT, false);
-        if (!vkd3d_create_vk_buffer_view(device, counter_resource->u.vk_buffer, format,
-                desc->u.Buffer.CounterOffsetInBytes + resource->heap_offset, sizeof(uint32_t), &view->vk_counter_view))
+        if (device->bindless_state.flags & VKD3D_BINDLESS_UAV_COUNTER)
         {
-            WARN("Failed to create counter buffer view.\n");
+            VkDeviceAddress address = vkd3d_get_buffer_device_address(device, counter_resource->u.vk_buffer);
+
             view->vk_counter_view = VK_NULL_HANDLE;
-            d3d12_desc_destroy(descriptor, device);
+            view->vk_counter_address = address + desc->u.Buffer.CounterOffsetInBytes;
+        }
+        else {
+            const struct vkd3d_format *format = vkd3d_get_format(device, DXGI_FORMAT_R32_UINT, false);
+
+            if (!vkd3d_create_vk_buffer_view(device, counter_resource->u.vk_buffer, format,
+                    desc->u.Buffer.CounterOffsetInBytes + resource->heap_offset, sizeof(uint32_t), &view->vk_counter_view))
+            {
+                WARN("Failed to create counter buffer view.\n");
+                view->vk_counter_view = VK_NULL_HANDLE;
+                d3d12_desc_destroy(descriptor, device);
+            }
+
+            view->vk_counter_address = 0;
```

Comment:
Should write 0 to address as well if we're binding a NULL counter.

Reply:
I'll just zero-initialize this in `vkd3d_view_create`, where we already set `vk_counter_view = VK_NULL_HANDLE` since we'd otherwise read undefined values in more cases.